pkg/i2gw/aggregator.go

```diff
@@ -0,0 +1,419 @@
+	ingressClass string
+	backend      networkingv1.IngressBackend
+}
+
+type ingressPath struct {
+	path  networkingv1.HTTPIngressPath
+	extra *extra
+}
+
+type extra struct {
+	canary *canary
+}
+
+type canary struct {
+	enable           bool
+	headerKey        string
+	headerValue      string
+	headerRegexMatch bool
+	weight           int
+	weightTotal      int
+}
+
+func (a *ingressAggregator) addIngress(ingress networkingv1.Ingress) {
+	var ingressClass string
+	if ingress.Spec.IngressClassName != nil && *ingress.Spec.IngressClassName != "" {
+		ingressClass = *ingress.Spec.IngressClassName
+	} else if _, ok := ingress.Annotations[networkingv1beta1.AnnotationIngressClass]; ok {
+		ingressClass = ingress.Annotations[networkingv1beta1.AnnotationIngressClass]
+	} else {
+		ingressClass = ingress.Name
+	}
+	e := getExtra(ingress)
+	for _, rule := range ingress.Spec.Rules {
+		a.addIngressRule(ingress.Namespace, ingressClass, rule, ingress.Spec, e)
+	}
+	if ingress.Spec.DefaultBackend != nil {
+		a.defaultBackends = append(a.defaultBackends, ingressDefaultBackend{
+			name:         ingress.Name,
+			namespace:    ingress.Namespace,
+			ingressClass: ingressClass,
+			backend:      *ingress.Spec.DefaultBackend,
+		})
+	}
+}
+
+func (a *ingressAggregator) addIngressRule(namespace, ingressClass string, rule networkingv1.IngressRule, iSpec networkingv1.IngressSpec, e *extra) {
+	rgKey := ruleGroupKey(fmt.Sprintf("%s/%s/%s", namespace, ingressClass, rule.Host))
+	rg, ok := a.ruleGroups[rgKey]
+	if !ok {
+		rg = &ingressRuleGroup{
+			namespace:    namespace,
+			ingressClass: ingressClass,
+			host:         rule.Host,
+		}
+		a.ruleGroups[rgKey] = rg
+	}
+	if len(iSpec.TLS) > 0 {
+		rg.tls = append(rg.tls, iSpec.TLS...)
+	}
+	rg.rules = append(rg.rules, ingressRule{rule: rule, extra: e})
+}
+
+func (a *ingressAggregator) toHTTPRoutesAndGateways() ([]gatewayv1beta1.HTTPRoute, []gatewayv1beta1.Gateway, []error) {
+	var httpRoutes []gatewayv1beta1.HTTPRoute
+	var errors []error
+	listenersByNamespacedGateway := map[string][]gatewayv1beta1.Listener{}
+
+	for _, rg := range a.ruleGroups {
+		listener := gatewayv1beta1.Listener{}
+		if rg.host != "" {
+			listener.Hostname = (*gatewayv1beta1.Hostname)(&rg.host)
+		} else if len(rg.tls) == 1 && len(rg.tls[0].Hosts) == 1 {
+			listener.Hostname = (*gatewayv1beta1.Hostname)(&rg.tls[0].Hosts[0])
+		}
+		if len(rg.tls) > 0 {
+			listener.TLS = &gatewayv1beta1.GatewayTLSConfig{}
+		}
+		for _, tls := range rg.tls {
+			listener.TLS.CertificateRefs = append(listener.TLS.CertificateRefs,
+				gatewayv1beta1.SecretObjectReference{Name: gatewayv1beta1.ObjectName(tls.SecretName)})
+		}
+		gwKey := fmt.Sprintf("%s/%s", rg.namespace, rg.ingressClass)
+		listenersByNamespacedGateway[gwKey] = append(listenersByNamespacedGateway[gwKey], listener)
+		httpRoute, errors := rg.toHTTPRoute()
+		httpRoutes = append(httpRoutes, httpRoute)
+		errors = append(errors, errors...)
+	}
+
+	for _, db := range a.defaultBackends {
```

Comment:
yeah i dont know what to do about a case where multiple ingresses have defined a default backend, since the Ingress spec is not so clear about it

error? split traffic among the various backends?